### DISCUSSION TITLE
Package manager for third party activities

### DISCUSSION THREAD

**Discussion by johnwc** [elsa-workflows/elsa-core#3158] [Ideas]
There should be a feature to allow a workflow server/dashboard to utilize a repository to search and install custom third party activities. I looked at the functionality of nuget search, and I believe it would be a good fit the repository type. Third party developers can publish their activities using a specific tag like `elsa-activities` or something similar. The dashboard can query the nuget repo using a tag filter to only pull back nuget packages with this tag. This also allows developers that want to pull down the activity into a custom Elsa .Net project, to be able to embed them in the project build.

I created this discussion to discuss if you, @sfmskywalker, view this as a viable feature, and if so to discuss the architect. I would be willing to contribute the development of this to the project. This all stemmed from us developing a few handful of activities that we think would be useful for others to use.

**Comment by sfmskywalker** [comment]
I love the idea, and I have been thinking of a package gallery as well.

Everything you describe makes sense, but I am hesitant to allowing the user to install packages at runtime into the application.
On the one hand, this is great during development. But on the other hand, you probably do not want to do this in production environments where you can potentially break the entire application. It would also be problematic in environments  that use automated build pipelines that build & deploy applications from source code repositories.

Regardless, having a package browser either as a separate gallery website and/or accessible straight from the dashboard would be sweet!

In addition, I am contemplating ways to make it super-easy for activity developers who want to share their packages by simply letting them register their repo with some service we control. This service then monitors commits to the main branch of the repo and automatically builds & deploys to NuGet (and MyGet for preview builds).

**Reply by johnwc** [reply]
I don't see the issue with a workflows admin being able to pull third party activities from a nuget repo to use them in a production environment. It would be the responsibility of the workflow admin to test them in dev first, and you can put a warning about being sure to test them in dev first. The only thing that will be run from the libraries would be the activities, and if the activity fails they can just not use them or remove them. This decision needs to be left up to the workflow admin if they want to use the library or not. 

How it would work: 
 - Elsa would pull down the pre-compiled libraries from nuget and extract them.
 - It would load the library(dll) and search for a class that inherits the `Elsa.Services.IActivity` and run `AddActivitiesFrom()`
   All this would of course be done within safe try/catch in case there was an exception from the library. 
 - The nuget would be registered in the persisted store, and upon startup Elsa would pull from nuget if not already download and load any libraries, allowing them to be used in workflows. 

I've used this plugin architect in the past, and it works quite well.

I don't quite understand where it would be problematic for the automatic build pipeline scenario, can you explain where you would see an issue? Since these libraries are in nuget, they can be baked right into the project as nuget references with the developer calling `AddActivitiesFrom()` manually.
